desc/optimize/tr_subproblems.py

```diff
@@ -370,6 +376,118 @@ def loop_body(state):
+        Initial guess for alpha, which might be available from a previous
+        iteration. If None, determined automatically.
+    rtol : float, optional
+        Stopping tolerance for the root-finding procedure. Namely, the
+        solution ``p`` will satisfy
+        ``abs(norm(p) - trust_radius) < rtol * trust_radius``.
+    max_iter : int, optional
+        Maximum allowed number of iterations for the root-finding procedure.
+
+    Returns
+    -------
+    p : ndarray, shape (n,)
+        Found solution of a trust-region problem.
+    hits_boundary : bool
+        True if the proposed step is on the boundary of the trust region.
+    alpha : float
+        Positive value such that (J.T*J + alpha*I)*p = -J.T*f.
+        Sometimes called Levenberg-Marquardt parameter.
+
+    """
+    # try full newton step
+    tall = J.shape[0] >= J.shape[1]
+    if tall:
+        Q, R = qr(J, mode="economic")
+        p_newton = solve_triangular_regularized(R, -Q.T @ f)
+    else:
+        Q, R = qr(J.T, mode="economic")
+        p_newton = Q @ solve_triangular_regularized(R.T, f, lower=True)
+
+    def truefun(*_):
+        return p_newton, False, 0.0
+
+    def falsefun(*_):
+        alpha_upper = jnp.linalg.norm(J.T @ f) / trust_radius
+        alpha_lower = 0.0
+        alpha = setdefault(
+            initial_alpha,
+            jnp.maximum(0.001 * alpha_upper, (alpha_lower * alpha_upper) ** 0.5),
+        )
+        k = 0
+        # algorithm 4.3 from Nocedal & Wright
+        fp = jnp.pad(f, (0, J.shape[1]))
+
+        def loop_cond(state):
+            alpha, alpha_lower, alpha_upper, phi, k = state
+            return (jnp.abs(phi) > rtol * trust_radius) & (k < max_iter)
+
+        def loop_body(state):
+            alpha, alpha_lower, alpha_upper, phi, k = state
+
+            alpha = jnp.where(
+                (alpha < alpha_lower) | (alpha > alpha_upper),
+                jnp.maximum(0.001 * alpha_upper, (alpha_lower * alpha_upper) ** 0.5),
+                alpha,
+            )
+
+            Ji = jnp.vstack([J, jnp.sqrt(alpha) * jnp.eye(J.shape[1])])
+            # Ji is always tall since its padded by alpha*I
+            Q, R = qr(Ji, mode="economic")
+            p = solve_triangular_regularized(R, -Q.T @ fp)
+            p_norm = jnp.linalg.norm(p)
+            phi = p_norm - trust_radius
+            alpha_upper = jnp.where(phi < 0, alpha, alpha_upper)
+            alpha_lower = jnp.where(phi > 0, alpha, alpha_lower)
+
+            q = solve_triangular_regularized(R.T, p, lower=True)
+            q_norm = jnp.linalg.norm(q)
+
+            alpha += (p_norm / q_norm) ** 2 * phi / trust_radius
+            alpha = jnp.where(
+                (alpha < alpha_lower) | (alpha > alpha_upper),
+                jnp.maximum(0.001 * alpha_upper, (alpha_lower * alpha_upper) ** 0.5),
+                alpha,
+            )
+            k += 1
+            return alpha, alpha_lower, alpha_upper, phi, k
+
+        alpha, *_ = while_loop(
```

Comment:
A single QR is faster than SVD on both CPU and GPU, but this `while_loop` is very slow. So, if we ever need to call `falsefun` that iteration is significantly slower than SVD on CPU. For GPU, I guess the execution of this `while_loop` is not that slow so we saw a speed-up with QR (not single QR but overall solution time decreased). Is there a better way to do this part?

Reply:
Not really if you want the exact solution to the subproblem. You can get an approximate solution using the dogleg or subspace method but I've found those don't perform as well.

Reply:
I will look into ways of optimizing how to decompose `Ji` given that you know decomposition J=QR. I am not sure but there should be some linear algebra trick we can use

Reply:
There are methods for updating a QR factorization when you add rows. Suppose we have

$$
QR = J
$$

what we want is 

$$
\tilde{Q} \tilde{R} = \begin{pmatrix} J \\ 
\alpha I \end{pmatrix}
$$

for different values of $\alpha$. scipy has functions to do this, though jax doesn't so we'd need to implement ourselves. The ones I'm familiar with are sequential (ie, adding 1 row at a time) so probably aren't great on GPU but there are probably smarter ways. You can look into damped least squares, updating qr factorizations etc.